tests/cache_test_state.rs

```diff
@@ -0,0 +1,81 @@
+use std::env;
+use std::sync::Arc;
+use std::time::Duration;
+
+use once_cell::sync::Lazy;
+use tokio::sync::watch::channel;
+
+use momento::config::configurations;
+use momento::{CacheClient, CredentialProviderBuilder, MomentoError};
+
+pub static TEST_STATE: Lazy<Arc<TestState>> = Lazy::new(|| Arc::new(TestState::new()));
+
+pub struct TestState {
+    pub client: Arc<CacheClient>,
+    pub cache_name: String,
+    #[allow(dead_code)]
+    runtime: tokio::runtime::Runtime,
+}
+
+impl TestState {
+    fn new() -> Self {
+        let cache_name =
+            env::var("TEST_CACHE_NAME").unwrap_or_else(|_| "rust-test-cache".to_string());
+        println!("Using cache name: {}", cache_name);
+        let thread_cache_name = cache_name.clone();
+
+        let credential_provider =
+            CredentialProviderBuilder::from_environment_variable("TEST_API_KEY".to_string())
+                .build()
+                .expect("auth token should be valid");
+
+        // The cache client must be created using a separate tokio runtime because each test
+        // creates it own runtime, and the client will stop running if its runtime is destroyed.
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .expect("test state tokio runtime failure");
+        let (sender, client_receiver) = channel(None);
+        let barrier = Arc::new(std::sync::Barrier::new(2));
+        let thread_barrier = barrier.clone();
+        runtime.spawn(async move {
+            let cache_client = CacheClient::new(
+                credential_provider,
+                configurations::laptop::latest(),
+                Duration::from_secs(5),
+            )
+            .expect("Failed to create cache client");
+
+            match cache_client.clone().create_cache(thread_cache_name).await {
+                Ok(_) => {}
+                Err(e) => match e {
+                    MomentoError::AlreadyExists { .. } => {
+                        println!("Cache already exists.");
+                    }
+                    _ => {
+                        panic!("Failed to create cache: {:?}", e);
+                    }
+                },
+            }
+
+            sender
+                .send(Some(cache_client))
+                .expect("client should be sent to test state thread");
+            thread_barrier.wait();
+        });
+        barrier.wait();
+
+        // Retrieve the client from the runtime that created it.
+        let client = client_receiver
```

Comment:
wow, fancy. I understand how this is working but am a little surprised that we have to use a channel to pass it back? naively, it seems like with an Arc we should be able to just reference the same client from multiple threads, but maybe Arcs are just for referencing from multiple functions in the same thread, and channels are required to move across thread boundaries?

anyway, I assume you and kenny paired on this and aligned on this solution, in which case I'm +1, just musing about how it reads to a newb :)

Reply:
Yeah, Kenny and I went through several frustrating steps to get to this point. There may be an easier solution, but I don't currently have the Rust chops to point it out.

Reply:
ack